server/middleware/passcodeStrategy.ts

```diff
@@ -0,0 +1,148 @@
+import { pgPool } from 'common/pool';
+import { PoolClient } from 'pg';
+import { Strategy as CustomStrategy } from 'passport-custom';
+import {
+  InvalidLoginParameter,
+  UnauthorizedError,
+  NotFoundError,
+} from '../errors';
+import ms from 'ms';
+
+export function PasscodeStrategy(): CustomStrategy {
+  return new CustomStrategy(async function (req, done) {
+    let client: PoolClient;
+    try {
+      client = await pgPool.connect();
+      const { email, passcode } = req.body;
+      const accountId = verifyPasscode({ email, passcode, client });
+      if (accountId) {
+        return done(null, { accountId });
+      }
+    } catch (err) {
+      return done(err);
+    } finally {
+      if (client) {
+        client.release();
+      }
+    }
+  });
+}
+
+type CreatePasscodeParams = {
+  email?: string;
+  client?: PoolClient;
+};
+
+/**
+ * Deletes unconsumed passcodes and create a new passcode for the given email
+ * @param createPasscodeParams Params for createPasscode function
+ * @param createPasscodeParams.email The email of the user requesting a passcode to sign in
+ * @param createPasscodeParams.client The pg PoolClient
+ * @returns Promise<passcode>
+ */
+export async function createPasscode({ email, client }: CreatePasscodeParams) {
+  try {
+    if (!email) {
+      throw new InvalidLoginParameter('Invalid email parameter');
+    }
```

Comment:
```suggestion
    }
    if (!client) {
      throw new Error('client is falsy');
    }     
```
not sure this is the best way to deal with this. but i was seeing a typescript warning in my editor:
![Screen Shot 2023-10-19 at 3 32 53 PM](https://github.com/regen-network/regen-server/assets/10120306/bae2cabd-b932-4a76-a172-1440c36376ee)

Reply:
maybe CreatePasscodeParams could just mark email and client as required instead, that's better but maybe there's a reason these are optional?

Reply:
similar thing applies in the verifyPasscode function

Reply:
it could be that the request doesn't have any email or passcode so I wanted to include this check as part of the functions themselves but I guess I could make client required